### ISSUE TITLE
Keyboard just shows "Insert symbol", "Done", and "Cycle Mode"

### ISSUE THREAD

**Issue by wongjustin98** [Clam-/TraditionalT9#17]
I'm stuck at my phone's lock screen since I can't click confirm after typing in my pass code. 

Pressing the center key on my phone just hits "done" and exits out of the keyboard prompt. 

Is there something I'm doing wrong? It seems like something on this keyboard maps incorrectly to my phone. I am using a SH-01J. 

I will try building this and messing with it on an emulator, but looking for help here in the meantime. I am running Android 5.X, not sure which version right now since I'm locked out of my phone.

**Comment by registered99** [comment]
Was able to get out using safe mode (On an SH-01J, that ended up being, holding the side button down while booting). 

![image](https://cloud.githubusercontent.com/assets/10359784/24177309/fd50aef8-0e77-11e7-95d3-d07490963dcd.png)

**Comment by Clam-** [comment]
Hmm sounds strange. I've had some occurrences when the IME window has popped up when I wasn't expecting it. Sometimes I've had to hardcode some input-types to never show the IME helper window.

Can you explain to me a little more about the lock screen issue you faced? Is it a standard lock screen? Am I able to test with an emulator on the emulator lock screen?
I don't actually use the lock screen on my phone, but I can try. I have a 007SH J though so it might act a little differently.

**Comment by Clam-** [comment]
Oh by the way, sorry the UI looks pretty janky on your screenshot. I've basically only designed it with my phone's DPI and layout in mind. 
It's weird that you have the native button overlay still showing. I also have a SH31 that has a similar overlay that I can test with. I've been meaning to transition to that phone and redesign the layout, but hacking Google Play in to that phone is a pain in the butt and it's put me off.

**Comment by wongjustin99** [comment]
I was able to get it to come up; I had to "Load Dictionary". Once that was done, I could start using it. 

It's the lock screen faced by having a PIN code on the phone. The reason it was broke was because I had a PIN password but my keyboard was switched to here. Since the IME wouldn't come up, I could type in the numbers for my password but was stuck. It's a standard screen on the Galapagos-based phones, but I'm not sure what 007sh might be running. The SH-01J runs something based off Android 5.1. 

I also was able to get the build running, so hopefully I'll be able to fix the UI sizing on my screen. No problem at all -- it gets expensive to get devices to test these on. 

Would you mind talking about how you have your emulator set up? I'm having an issue getting the hardware keys working. My phone also doesn't seem to have the same soft keys you have, so I can't use them for those functions.

**Comment by wongjustin99** [comment]
One more question if it's not too much trouble: Is it possible to change the order of words in the dictionary? There seems to be a preference for acronyms (HIV coming up before got, etc). Also it's hard to type single-letter words common in english such as 'I' or 'a' without getting the European-variants on top.

**Comment by Clam-** [comment]
When you say "I was able to get it to come up", what do you mean? Yes you need to load a dictionary first before any suggestions can be displayed as you type.

Words have a "priority" or "ranking" assigned to them. Each time you use a word, that increases. So eventually your most common words will rise above others.
I explored the option of feeding usage data to seed the ranking, but it looked to tedious so I didn't do it.
If you wanted to fudge with it though, if you backup your current dictionary, you'll see how the ranking is structured and you can just mess with that txt file and load the backup and it'll store the ranking you've modified.

With the emulator setup, you have to modify the configuration files to enable hardware keyboard/keypad. I'll have a look when I get home and show you the things you need to change.

The first time you commit a single letter word, that will appear at the top of the list the next time you go to enter that word (unless you commonly type in European single letter words).
Hope this makes sense. Well that's the way it's supposed to work.

**Comment by Clam-** [comment]
I just had a look at my emulator configuration and it contains these settings that are probably most relevant
```
hw.dPad = true
hw.keyboard=yes
```
I also use this keymap to rebind keys I think to the numpad to make it easier for testing:
https://github.com/Clam-/TraditionalT9/blob/master/keymap.txt.emulator

I haven't played with this stuff in some time so it might take some experimentation. My emulator doesn't seem to be starting up so I'll have to fix that sometime.

**Comment by wongjustin98** [comment]
"I was able to get it to come up" was for getting the IME window to show. I think I didn't understand that you had to first pre-load the dictionary before it would be usable. I guess I assumed some kind of fallback in case a dictionary didn't exist. 

I do see it becoming a little better after some usage. Maybe I'll see if there's another T9 dictionary that I can add with pre-seeded frequency. 

Ah, I've narrowed it down to the key events on my emulator not matching up to SOFT_LEFT and SOFT_RIGHT. However, I can emulate those on my phone well-enough by doing ​​​`adb shell input keyevent 1` and  ​​​`adb shell input keyevent 2`.
